protocol/contact_test.go

```diff
@@ -559,3 +560,152 @@ func TestMarshalContactJSON(t *testing.T) {
 	require.True(t, strings.Contains(string(encodedContact), "compressedKey\":\"zQ"))
 
 }
+
+func TestContactContactRequestPropagatedStateReceivedOutOfDateLocalStateOnTheirSide(t *testing.T) {
+	// We receive a message with expected contact request state != our state
+	// and clock < our clock, we ping back the user to reach consistency
+
+	c := &Contact{}
+	c.ContactRequestLocalState = ContactRequestStateSent
+	c.ContactRequestLocalClock = 1
+
+	result := c.ContactRequestPropagatedStateReceived(
+		&protobuf.ContactRequestPropagatedState{
+			RemoteState: uint64(ContactRequestStateNone),
+			RemoteClock: 0,
+			LocalState:  uint64(ContactRequestStateNone),
+			LocalClock:  1,
+		},
+	)
+
+	require.True(t, result.sendBackState)
+
+	// if the state is the same, it should not send back a message
+
+	c = &Contact{}
+	c.ContactRequestLocalState = ContactRequestStateNone
+	c.ContactRequestLocalClock = 1
+
+	result = c.ContactRequestPropagatedStateReceived(
+		&protobuf.ContactRequestPropagatedState{
+			RemoteState: uint64(ContactRequestStateNone),
+			RemoteClock: 0,
+			LocalState:  uint64(ContactRequestStateNone),
+			LocalClock:  1,
+		},
+	)
+
+	require.False(t, result.sendBackState)
+
+	// If the clock is the same, it should not send back a message
+	c = &Contact{}
+	c.ContactRequestLocalState = ContactRequestStateSent
+	c.ContactRequestLocalClock = 1
+
+	result = c.ContactRequestPropagatedStateReceived(
+		&protobuf.ContactRequestPropagatedState{
+			RemoteState: uint64(ContactRequestStateNone),
+			RemoteClock: 1,
+			LocalState:  uint64(ContactRequestStateNone),
+			LocalClock:  1,
+		},
+	)
+
+	require.False(t, result.sendBackState)
+
+}
+
+func TestContactContactRequestPropagatedStateReceivedOutOfDateLocalStateOnOurSide(t *testing.T) {
+	// We receive a message with expected contact request state == none
+	// and clock > our clock. We consider this a retraction, unless we are in the dismissed state, since that should be only changed by a trusted device
+
+	c := &Contact{}
+	c.ContactRequestLocalState = ContactRequestStateSent
+	c.ContactRequestLocalClock = 1
+
+	c.ContactRequestPropagatedStateReceived(
+		&protobuf.ContactRequestPropagatedState{
+			RemoteState: uint64(ContactRequestStateNone),
+			RemoteClock: 2,
+			LocalState:  uint64(ContactRequestStateNone),
+			LocalClock:  1,
+		},
+	)
+
+	require.False(t, c.added())
```

Comment:
The assertions in `TestContactContactRequestPropagatedStateReceivedOutOfDateLocalStateOnTheirSide` verify the result returned by `ContactRequestPropagatedStateReceived`, but these ones only check the mutations in `c`. I think these tests would give stronger guarantees if they also asserted the full returned result from `ContactRequestPropagatedStateReceived`, as well as the mutations on `c`.

Reply:
I'll fortify the tests in a follow up PR, if that's ok, I have yet one fix to add separately :)